data/transactions/logic/assembler.go

```diff
@@ -1189,6 +1187,197 @@ func (ops *OpStream) checkArgs(spec OpSpec) {
+						err := fmt.Errorf("different paths down nearby loop cause different loop top stacks")
+						ops.warnf("%w", err)
+						//This should also cover the case where two loops have the same head but change the stack in different ways
+					}
+					//else case means all good, but we should not keep going through blocks as we already sent a path down before
+				}
+			}
+		}
+	}
+}
+
+func (ops *OpStream) annotateBlockInOut() {
+	blocks := ops.blocks
+	var j int
+	var spec OpSpec
+	var top []StackType
+	var out []StackType
+	raw := ops.pending.Bytes()
+	for i := range blocks {
+		j = blocks[i].startIndex
+		ops.typeStack = ops.typeStack[:0]
+		top = make([]StackType, 0)
+		for j <= blocks[i].endIndex {
+			spec = opsByOpcode[ops.Version][raw[j]]
+			for k := len(spec.Args) - 1; k >= 0; k-- {
+				argType := spec.Args[i]
+				sType := ops.tpop()
+				if sType == StackNone {
+					top = append(top, argType)
+				} else if (argType != StackAny) && (sType != StackAny) && (argType != sType) {
+					ops.sourceLine = ops.OffsetToLine[j]
+					err := fmt.Errorf("%s arg %d wanted type %s got %s", spec.Name, i, argType.String(), sType.String())
+					ops.error(err)
+				}
+			}
+			if len(spec.Returns) > 0 {
+				ops.tpusha(spec.Returns)
+			}
+			j = j + spec.Details.Size
+		}
+		//Not sure if slices require you to make copies like this
+		out = make([]StackType, len(ops.typeStack))
+		copy(out, ops.typeStack)
+		blocks[i].stackDelta = out
+		blocks[i].expectedStackTop = top
+
+	}
+}
+func (ops *OpStream) addBlockJumps(indexToBlock map[int]int) {
+	blocks := ops.blocks
+	raw := ops.pending.Bytes()
+	var spec OpSpec
+	for i := range blocks {
+		spec = opsByOpcode[ops.Version][raw[blocks[i].endIndex]]
+		switch spec.Name {
+		case "bz", "bnz", "b":
+			x, ok := indexToBlock[blocks[i].jumpTo]
+			if ok {
+				blocks[i].jumpTo = x
+			} else {
+				blocks[i].jumpTo = exiting //if a label index is not in indexToBlock it is an end label or super weird
+			}
+		}
+	}
+}
+
+//createBlocks sets up the basic blocks to be used in static analysis; note no block is added for labels at end of program
+func (ops *OpStream) createBlocks(reverseLabels map[int]string) map[int]int {
+	indexToBlock := make(map[int]int)
+	var blocks []BasicBlock
+	raw := ops.pending.Bytes()
+	startI := 0
+	i := 0         //index over raw
+	refsIndex := 0 //taking advantage of fact labelRefs is "sorted" automatically
+	var spec OpSpec
+	var _ string
+	var ok bool
+	for i < len(raw) {
+		spec = opsByOpcode[ops.Version][raw[i]]
+		_, ok = reverseLabels[i+spec.Details.Size] //checks if next instruction was a label
+		if i == ops.labelReferences[refsIndex].position {
+			//We put the start indices of where we jump to b/c we will have map from start indices to blocks after
+			if spec.Name == "bz" || spec.Name == "bnz" {
+				blocks = append(blocks, BasicBlock{startIndex: startI, endIndex: i, jumpTo: ops.labels[ops.labelReferences[refsIndex].label], flowTo: len(blocks) + 1})
```

Comment:
Perhaps these four big, similar `append` lines can be made clearer by having an ops.addBlock() method.  So you would stop operating on a local `blocks` slice and directly express that you're adding to ops.blocks with such a call.